### PR TITLE
DAOS-9200 Remove compatibility code for daos_obj_generate_oid

### PR DESCRIPTION
**Pull request by wiliamhuang**
daos_obj_generate_oid() is removed. Only daos_obj_generate_oid2 is
supported in API. daos_obj_generate_oid() is used in all other
places. Type daos_ofeat_t is removed. Only enum daos_otype_t is
used.

Signed-off-by: Lei Huang <lei.huang@intel.com>

### CODE REVIEW

File: src/vos/tests/vts_io.h [hunk 1]
```diff
@@ -71,7 +71,7 @@ struct io_test_args {
 	const char		*dkey;
 	const char		*akey;
 	void			*custom;
-	int			 ofeat;
+	enum daos_otype_t 	 ofeat;
```
[the remark this answers is not in the document]
Comment (wiliamhuang) [reply src/vos/tests/vts_io.h:74]: I just followed the format in this file. Please let me know if you do want me to remove the one space around this line. Thank you!

File: src/vos/tests/vts_io.c [hunk 2]
```diff
@@ -2549,32 +2549,32 @@ run_io_test(daos_ofeat_t feats, int keys, bool nest_iterators, const char *cfg)
 
 	init_num_keys = VTS_IO_KEYS;
 
-	feats = feats & DAOS_OF_MASK;
-	if ((feats & DAOS_OF_DKEY_UINT64) && (feats & DAOS_OF_DKEY_LEXICAL)) {
+	type = type & DAOS_OF_MASK;
+	if ((type & DAOS_OF_DKEY_UINT64) && (type & DAOS_OF_DKEY_LEXICAL)) {
```
Comment (bfaccini) [inline src/vos/tests/vts_io.c:2553]: Don't you need to better use the "..._OT..." versions of defines ?

Comment (wiliamhuang) [reply src/vos/tests/vts_io.c:2553]: Thank you! Yes. I will use *_OT_* and remove _OT_ version definition. I also need to add "DAOS_OT_MASK" to replace "DAOS_OF_MASK".

**Review by jolivier23 — CHANGES_REQUESTED** [review]
API_VERSION_MAJOR = "3" ?

**Comment by wiliamhuang** [comment]
> API_VERSION_MAJOR

I did not change API_VERSION_MAJOR. Where do you see this?

**Comment by wiliamhuang** [comment]
hdf5_vol_daos was installed in CI from prebuilt rpm, so it is not aware of the changes in daos head files and is still looking for daos_obj_generate_oid(). This caused many failures in "Test Hardware / Functional Hardware Small / FTEST_daos_vol.DaosVol".

**Comment by wiliamhuang** [comment]
Now both daos_obj_generate_oid() and daos_obj_generate_oid2() are exposed in binary. daos_obj_generate_oid2 is a weak symbol pointing to daos_obj_generate_oid().

File: src/vos/tests/vts_pm.c [hunk 3]
```diff
@@ -2758,13 +2758,13 @@ test_multiple_key_conditionals_common(void **state, bool with_dtx)
...
 	/* set up dkey and akey */
-	oid = gen_oid(arg->ofeat);
+	oid = gen_oid(arg->otype);
 	vts_key_gen(&dkey_buf[0], arg->dkey_size, true, arg);
 	vts_key_gen(&akey1_buf[0], arg->akey_size, false, arg);
 	vts_key_gen(&akey2_buf[0], arg->akey_size, false, arg);
-	set_iov(&dkey, &dkey_buf[0], arg->ofeat & DAOS_OF_DKEY_UINT64);
-	set_iov(&akey1, &akey1_buf[0], arg->ofeat & DAOS_OF_AKEY_UINT64);
-	set_iov(&akey2, &akey2_buf[0], arg->ofeat & DAOS_OF_AKEY_UINT64);
+	set_iov(&dkey, &dkey_buf[0], arg->otype & DAOS_OT_DKEY_UINT64);
```
Comment (jolivier23) [inline src/vos/tests/vts_pm.c:2765]: these are not bit masks, shouldn't this be == ?

**Review by jolivier23 — CHANGES_REQUESTED** [review]

**Comment by wiliamhuang** [comment]
Removed DAOS_OF_* related bit ops. Waiting for the update of daos_vol in CI.

**Comment by wiliamhuang** [comment]
All CI tests passed. Could you please review this PR? Thank you!

File: src/common/btree.c [hunk 4]
```diff
@@ -4164,7 +4164,7 @@ dbtree_class_register(unsigned int tree_class, uint64_t tree_feats,
 }
 
 int
-dbtree_overhead_get(int alloc_overhead, unsigned int tclass, uint64_t ofeat,
+dbtree_overhead_get(int alloc_overhead, unsigned int tclass, uint64_t otype,
```
Comment (wangshilong) [inline src/common/btree.c:4167]: Maybe use daos_otype_t otype as well.

File: src/include/daos_obj.h [hunk 5]
```diff
@@ -91,8 +93,11 @@ enum daos_otype_t {
 
 	/** Byte Array with no metadata (eg DFS/POSIX) */
 	DAOS_OT_ARRAY_BYTE	= 13,
+	/** reserved: 1-level flat KV store */
+	DAOS_OT_KV_FLAT		= 14,
+
+	DAOS_OT_MAX		= 14,
```
Comment (wangshilong) [inline src/include/daos_obj.h:99]: [defect]This should be 15?

File: src/include/daos_srv/vos.h [hunk 6]
```diff
@@ -1075,14 +1075,14 @@ vos_obj_query_key(daos_handle_t coh, daos_unit_oid_t oid, uint32_t flags,
...
- *  \param ofeat[IN]		Relevant object features
+ *  \param otype[IN]		Relevant object features
  *  \param ovhd[IN,OUT]		Returned overheads
  *
  *  \return 0 on success, error otherwise.
  */
 int
 vos_tree_get_overhead(int alloc_overhead, enum VOS_TREE_CLASS tclass,
-		      uint64_t ofeat, struct daos_tree_overhead *ovhd);
+		      uint64_t otype, struct daos_tree_overhead *ovhd);
```
Comment (wangshilong) [inline src/include/daos_srv/vos.h:1085]: daos_otype_t

File: src/include/daos_obj.h (hunk 5 above)
Comment (wangshilong) [reply src/include/daos_obj.h:99]: I see, this is included.

**Review by wangshilong — APPROVED** [review]

File: src/vos/tests/vts_io.h [hunk 7]
```diff
@@ -71,7 +71,7 @@ struct io_test_args {
 	const char		*dkey;
 	const char		*akey;
 	void			*custom;
-	int			 ofeat;
+	enum daos_otype_t 	 otype;
```
[the remark this answers is not in the document]
Comment (jolivier23) [reply src/vos/tests/vts_io.h:74]: Looks like you have a space after daos_otype_t that can be removed.

**Review by jolivier23 — CHANGES_REQUESTED** [review]

File: src/vos/tests/vts_io.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (wiliamhuang) [reply src/vos/tests/vts_io.h:74]: The space is kept there to be consistent with the following three lines. They all have a space to align with the above lines with a * sign before variable names.

File: src/vos/tests/vts_io.c [hunk 8]
```diff
@@ -2634,32 +2637,22 @@ run_io_test(daos_ofeat_t feats, int keys, bool nest_iterators, const char *cfg)
...
 
-	feats = feats & DAOS_OF_MASK;
-	if ((feats & DAOS_OF_DKEY_UINT64) && (feats & DAOS_OF_DKEY_LEXICAL)) {
-		D_PRINT("Skipping ambiguous ofeat mask\n");
-		return 0;
-	}
-	if ((feats & DAOS_OF_AKEY_UINT64) && (feats & DAOS_OF_AKEY_LEXICAL)) {
-		D_PRINT("Skipping ambiguous ofeat mask\n");
-		return 0;
-	}
```
Comment (bfaccini) [inline src/vos/tests/vts_io.c:2645]: Do the same checks with corresponding DAOS_OT_... types no longer apply ??

**Review by bfaccini — APPROVED** [review]

File: src/vos/tests/vts_io.c (hunk 8 above)
Comment (wiliamhuang) [reply src/vos/tests/vts_io.c:2645]: Right. DAOS_OT_* does not support bit operation like DAOS_OF_*. The checks are not needed any more.

**Comment by wiliamhuang** [comment]
@wangshilong @bfaccini Could you please review? Due to the delay of the landing, I merged with master and fixed a couple of issues from the latest changes on May 6. Now all CI tests passed. I need to push this to land soon. Thank you very much!

**Review by wangshilong — APPROVED** [review]

**Review by bfaccini — APPROVED** [review]

**Comment by wiliamhuang** [comment]
@daos-stack/daos-gatekeeper Could you please review and land this PR soon? Thank you!

**Comment by wiliamhuang** [comment]
@wangshilong @bfaccini Just removed "DAOS_OT_KV_FLAT" which is not needed. "DAOS_OT_KV_HASHED" is used to replace "DAOS_OF_KV_FLAT". Could you please review again. Just need to check the commits made today. Thank you very much!

**Review by johannlombardi — APPROVED** [review]

**Review by wangshilong — APPROVED** [review]

**Pull request merged**